api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+
+### State Restoration For Drreg
+
+Owner: [Abhinav Sharma](https://github.com/abhinav92003)
+
+Upstream issue: [DynamoRIO/dynamorio#3823](https://github.com/DynamoRIO/dynamorio/issues/3823), [DynamoRIO/dynamorio#3801](https://github.com/DynamoRIO/dynamorio/issues/3801)
+
+On a state restore event, drreg should be able to restore all spilled registers to their application values.
+
+Unfortunately, when a state restore event happens, we only have the encoded fragment, and none of the drreg state, like the reg to spill slot mappings. We need to reconstruct this state based on the faulting pc and the encoded fragment.
+
+
+It is complex to determine which registers need to be restored and from which spill slot. This is because drreg automatically adds spill and restore instructions to handle various complex cases like automatic re-spilling of reserved registers after their application write instruction, and automatic restore of reserved registers before their application read instruction. Drreg also uses various optimisations like lazy restores for application value in case the register is reserved again. This is even more complex for aflags, for which spill and restore are a two-step process (spilling aflags involves reading aflags into a register using `lahf` and then writing that register to a spill slot; restoring aflags involves reading aflags from its spill slot to a register, and then writing aflags from that register using `sahf`). In some cases, aflags are even kept in a register as an optimisation.
+
+Additionally, in multi-phase use, a register may be spilled by multiple phases, with a separate spill slot for each phase. The application value for the register may reside in one or more spill slots, and may also move between spill slots based on how the spill regions from different phases overlap. See various tricky scenarios in [drreg-test.c](https://github.com/DynamoRIO/dynamorio/blob/f1d496b451eaa6e9aaff7125617030164c6cfdff/suite/tests/client-interface/drreg-test.c#L637).
+
+We explored two ways to adapt drreg’s state restoration logic to multi-phase use. This also fixed some known existing issues with drreg: [Dynamorio/dynamorio#4933](https://github.com/DynamoRIO/dynamorio/issues/4933), [DynamoRIO/dynamorio#4939](https://github.com/DynamoRIO/dynamorio/issues/4939).
```

Comment:
(Not requesting a change: just noting that I wouldn't normally expect to see references to past now-fixed bugs in a design doc.)